### PR TITLE
ENH: Allow map with abc mapping

### LINKED ISSUE #29733: Series.map(m: Mapping) can fail with "object is not callable"
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from collections import UserDict, Counter

class MissingDict(UserDict):
    def __missing__(self, key):
        return key

s = pd.Series(list("aab"))
s.map(Counter({"a": 1}))
# 0    1
# 1    1
# 2    0
# dtype: int64

s.map(MissingDict({"a": "c"}))
# <ipython-input-48-6cd4b0d364c9> in <module>
# ----> 1 s.map(MissingDict({"a": "c"}))

# /usr/local/lib/python3.7/site-packages/pandas/core/series.py in map(self, arg, na_action)
#    3826         dtype: object
#    3827         """
# -> 3828         new_values = super()._map_values(arg, na_action=na_action)
#    3829         return self._constructor(new_values, index=self.index).__finalize__(self)
#    3830 

# /usr/local/lib/python3.7/site-packages/pandas/core/base.py in _map_values(self, mapper, na_action)
#    1298 
#    1299         # mapper is a function
# -> 1300         new_values = map_f(values, mapper)
#    1301 
#    1302         return new_values

# pandas/_libs/lib.pyx in pandas._libs.lib.map_infer()

# TypeError: 'MissingDict' object is not callable
```

#### Problem description

`Series.map` can sometimes take Mapping subclasses other than `dict`, but not always. I think it would make sense if any object that satisfied `isinstance(x, collections.abc.Mapping)` could be used as a mapping. I came across this while trying to use map where I only wanted values present as keys in the mapping to change (I went for `s.map(lambda x: {"a": "b"}.get(x, x))` as a workaround).

`collections.Counter` and `collections.defaultdict` work fine, which made me expect other `Mapping`s would as well.

#### Expected Output

I expected `s.map(MissingDict({"a": "c"}))` to do something like:

```python
s.map(lambda x: MissingDict({"a": "c"})[x])
# 0    c
# 1    c
# 2    b
# dtype: object
```

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.17.4
pytz             : 2018.9
dateutil         : 2.7.5
pip              : 19.3.1
setuptools       : 41.6.0
Cython           : 0.29.14
pytest           : 5.2.2
hypothesis       : 4.44.1
sphinx           : 2.2.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.9.0
pandas_datareader: None
bs4              : 4.8.1
bottleneck       : None
fastparquet      : 0.3.2
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.0.3
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 2.6.2
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
s3fs             : None
scipy            : 1.3.2
sqlalchemy       : None
tables           : 3.6.1
xarray           : 0.14.0
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : None
```

</details>

### PR DESCRIPTION
**Pull request by ohad83**
- [x] closes #29733 (partly, look at my issue comment for details on why it won't exactly work on the issue's example)
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -559,6 +559,7 @@ Other
 - Bug in :meth:`DataFrame.append` that raised ``IndexError`` when appending with empty list (:issue:`28769`)
 - Fix :class:`AbstractHolidayCalendar` to return correct results for
   years after 2030 (now goes up to 2200) (:issue:`27790`)
+- Bug in :meth:`Series.map` where the mapper could not be a dict-like object which subclasses `collections.abc.Mapping` (:issue:`29733`)
```
[the remark this answers is not in the document]
Comment (ohad83) [reply doc/source/whatsnew/v1.0.0.rst]: Alright, moved it to enhancement.

File: pandas/core/series.py [hunk 2]
```diff
@@ -253,7 +254,7 @@ def __init__(
                 else:
                     data = data.reindex(index, copy=copy)
                 data = data._data
-            elif isinstance(data, dict):
+            elif isinstance(data, abc.Mapping):
```
Comment (jreback) [inline pandas/core/series.py]: this is an orthogonal change pls remove

File: pandas/core/series.py [hunk 3]
```diff
@@ -3617,7 +3618,7 @@ def map(self, arg, na_action=None):
 
         Parameters
         ----------
-        arg : function, dict, or Series
+        arg : function, dict, colletions.abc.Mapping subclass or Series
```
Comment (jreback) [inline pandas/core/series.py]: dict is now superfluous

File: pandas/tests/series/test_apply.py [hunk 4]
```diff
@@ -609,6 +609,51 @@ class DictWithoutMissing(dict):
         expected = Series([np.nan, np.nan, "three"])
         tm.assert_series_equal(result, expected)
 
+    def test_map_abc_mapping(self):
+        class NonDictMapping(abc.Mapping):
```
Comment (jreback) [inline pandas/tests/series/test_apply.py]: add the issue number as a comment

File: pandas/core/base.py [hunk 5]
```diff
@@ -1215,8 +1215,8 @@ def _map_values(self, mapper, na_action=None):
         # we can fastpath dict/Series to an efficient map
         # as we know that we are not going to have to yield
         # python types
-        if isinstance(mapper, dict):
-            if hasattr(mapper, "__missing__"):
+        if isinstance(mapper, abc.Mapping):
+            if isinstance(mapper, dict) and hasattr(mapper, "__missing__"):
```
Comment (jreback) [inline pandas/core/base.py]: why do you think you need the if condition that was added?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 5 above)
Comment (ohad83) [reply pandas/core/base.py]: `collections.abc.Mapping` does not declare nor implement `__missing__`, that is a concept of `dict`. In case you use a subclass of `collections.abc.Mapping` which isn't a subclass of `dict`, you cannot rely on `__missing__`.
You can see an example on the second test I added, `test_map_abc_mapping_with_missing`.

File: pandas/core/series.py (hunk 2 above)
Comment (ohad83) [reply pandas/core/series.py]: It is not orthogonal.
In the function `_map_values` in `core/base.py`, in case the mapper is a subclass of `collections.abc.Mapping` and isn't a `dict` with the `__missing__` property, we create a Series out of the mapper. In case it is `collections.abc.Mapping` and isn't a `dict`, the creation of the Series will fail if not for this change.

File: pandas/core/series.py (hunk 3 above)
Comment (ohad83) [reply pandas/core/series.py]: Wasn't sure if I should remove it as I suppose many standard users are familiar with `dict` but unfamiliar with `Mapping`.
Removed it now.

File: pandas/core/base.py (hunk 5 above)
Comment (WillAyd) [reply pandas/core/base.py]: I think can just simplify to `isinstance(mapper, dict)` then no? Check for __missing__ after that is duplicative?

File: pandas/core/series.py (hunk 2 above)
Comment (WillAyd) [reply pandas/core/series.py]: Is this hit in a test case?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks good generally. cc @TomAugspurger and @simonjayhawkins for related discussions that we've had on types - not sure if this is something we would need to review holistically

File: pandas/core/base.py (hunk 5 above)
Comment (ohad83) [reply pandas/core/base.py]: No, since `dict` doesn't have a `__missing__` method on its own.
In `dict.__getitem__`'s implementation there is a reference to `__missing__` in case the object is a subclass of `dict` and implements it, but `dict` itself does not implement `__missing__`.
I suggest reading [dict's documentation](https://docs.python.org/3/library/stdtypes.html#dict), specifically the `__getitem__` part noted as `d[key]` for explanation.

File: pandas/core/series.py (hunk 2 above)
Comment (ohad83) [reply pandas/core/series.py]: It is in the test cases I added. If I remove this change, they will fail, as a Series is created from the mapper.

**Comment by TomAugspurger** [comment]
Changes here look sensible.

It's not immediately clear to my why `Series.__init__` had to change. Can
we add a test in pandas/tests/series/test_constructors.py that shows the
change?

We may want to move these Mapping subclass test helpers to a central place
and reuse them.

**Comment by ohad83** [comment]
> Changes here look sensible. It's not immediately clear to my why `Series.__init__` had to change. Can we add a test in pandas/tests/series/test_constructors.py that shows the change? We may want to move these Mapping subclass test helpers to a central place and reuse them.

Added a test for constructing a `Series` with `NonDictMapping` and moved the `NonDictMapping` test subclass to `pandas/util/testing.py`.

File: pandas/core/base.py [hunk 6]
```diff
@@ -2,7 +2,7 @@
 Base and utility classes for pandas objects.
 """
 import builtins
-from collections import OrderedDict
+from collections import OrderedDict, abc
```
Comment (simonjayhawkins) [inline pandas/core/base.py]: minor nit. IMO shouldn't import abc from collections. to avoid confusion and potential future clashes with the abc module.

File: pandas/core/series.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 """
 Data structure for 1-dimensional cross-sectional and time series data
 """
+from collections import abc
```
Comment (simonjayhawkins) [inline pandas/core/series.py]: same

File: pandas/core/series.py [hunk 8]
```diff
@@ -3617,7 +3618,7 @@ def map(self, arg, na_action=None):
 
         Parameters
         ----------
-        arg : function, dict, or Series
+        arg : function, colletions.abc.Mapping subclass or Series
```
Comment (simonjayhawkins) [inline pandas/core/series.py]: I think could simply just use "mapping" in docstrings

File: pandas/core/base.py (hunk 6 above)
Comment (ohad83) [reply pandas/core/base.py]: I agree, though there are other files in the code which do that so I figured it's fine.
Changed.

File: pandas/core/series.py (hunk 8 above)
Comment (ohad83) [reply pandas/core/series.py]: I am not sure about that, the other docstrings I've seen in pandas use the whole `collections.abc.Mapping` and I believe it's better for the users who aren't familiar with `collections.abc`. Having the whole path can help them search it, while just `Mapping` won't be as helpful.
You did make me notice I had a typo though 😅

File: pandas/core/base.py (hunk 6 above)
Comment (ohad83) [reply pandas/core/base.py]: Apparently the guidelines in code_check.sh point to using `from collections import abc` instead of `from collections.abc import`, so changes reverted.

Comment (simonjayhawkins) [reply pandas/core/base.py]: fair enough. not sure of the reason for that.

File: pandas/core/series.py (hunk 2 above)
Comment (jreback) [inline pandas/core/series.py]: you can use is_dict_like here instead

File: pandas/core/base.py [hunk 9]
```diff
@@ -1215,8 +1215,8 @@ def _map_values(self, mapper, na_action=None):
         # we can fastpath dict/Series to an efficient map
         # as we know that we are not going to have to yield
         # python types
-        if isinstance(mapper, dict):
-            if hasattr(mapper, "__missing__"):
+        if isinstance(mapper, abc.Mapping):
```
Comment (jreback) [inline pandas/core/base.py]: use is_dict_like

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Review by WillAyd — APPROVED** [review]
lgtm. @ohad83 can you merge master one more time? I think should get CI green

**Comment by ohad83** [comment]
@WillAyd Merged but still a test seems to fail, I am not really sure what's happening and how, if at all, it's related to this PR.. Any ideas?

**Comment by ohad83** [comment]
> lgtm. @ohad83 can you merge master one more time? I think should get CI green

@jreback @WillAyd It's green now

File: pandas/util/testing.py [hunk 10]
```diff
@@ -2121,6 +2121,20 @@ def __init__(self, *args, **kwargs):
         dict.__init__(self, *args, **kwargs)
 
 
+class TestNonDictMapping(abc.Mapping):
```
Comment (WillAyd) [inline pandas/util/testing.py]: Can you make this a fixture in `pandas/tests/series/conftest.py` instead of putting in `pandas.util.testing`? With the latter sometimes things can leak into our API and definitely wouldn't want that here

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
@ohad83 can you rebase and address Will's comment if still relevant

**Comment by jreback** [comment]
can you merge master and will look again

**Comment by jreback** [comment]
@ohad83 going to have this close the original issue. if the remaining case is worthile, then pls open a new issue.

**Comment by jreback** [comment]
@WillAyd if any comments

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @ohad83 great PR
